File: lib/domain/agent/tool_preference_tracker.dart
class ToolPreferenceTracker {
  ToolPreferenceTracker({int topN = 3}) : _topN = topN;

  final int _topN;
  final Map<String, int> _usage = {};

  int get totalUsage => _usage.values.fold(0, (sum, count) => sum + count);

  int getCount(String toolName) => _usage[toolName] ?? 0;

  void recordToolUse(String toolName) {
    _usage[toolName] = (_usage[toolName] ?? 0) + 1;
  }

  List<String> getMostUsed([int? count]) {
    final sorted = _usage.entries.toList()
      ..sort((a, b) => b.value.compareTo(a.value));
    final limit = count ?? sorted.length;
    return sorted.take(limit).map((e) => e.key).toList();
  }

  String toPromptContext() {
    final top = getMostUsed(_topN);
    if (top.isEmpty) return '';
    final buffer = StringBuffer()..writeln('FREQUENTLY USED TOOLS:');
    for (final tool in top) {
      buffer.writeln('- $tool (${_usage[tool]} uses)');
    }
    return buffer.toString();
  }

  void clear() {
    _usage.clear();
  }
}

File: lib/src/agent/tool-preference-tracker.ts
export class ToolPreferenceTracker {
  private usage = new Map<string, number>();

  constructor(private topN = 3) {}

  recordToolUse(toolName: string): void {
    this.usage.set(toolName, (this.usage.get(toolName) ?? 0) + 1);
  }

  toPromptContext(): string {
    const top = this.getMostUsed(this.topN);
    if (top.length === 0) return '';
    const lines: string[] = ['FREQUENTLY USED TOOLS:'];
    for (const tool of top) {
      lines.push(`- ${tool} (${this.usage.get(tool)} uses)`);
    }
    return lines.join('\n');
  }

  private getMostUsed(count?: number): string[] {
    const sorted = [...this.usage.entries()].sort((a, b) => b[1] - a[1]);
    return sorted.slice(0, count ?? sorted.length).map(([k]) => k);
  }

  clear(): void {
    this.usage.clear();
  }
}
